// @opengsn/contracts/types/ethers-contracts/IStakeManager.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumber,
  BigNumberish,
  BytesLike,
  CallOverrides,
  ContractTransaction,
  Overrides,
  PopulatedTransaction,
  Signer,
  utils,
} from "ethers";
import type {
  FunctionFragment,
  Result,
  EventFragment,
} from "@ethersproject/abi";
import type { Listener, Provider } from "@ethersproject/providers";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
} from "./common";

export declare namespace IStakeManager {
  export type StakeInfoStruct = {
    stake: BigNumberish;
    unstakeDelay: BigNumberish;
    withdrawTime: BigNumberish;
    abandonedTime: BigNumberish;
    keepaliveTime: BigNumberish;
    token: string;
    owner: string;
  };

  export type StakeInfoStructOutput = [
    BigNumber,
    BigNumber,
    BigNumber,
    BigNumber,
    BigNumber,
    string,
    string
  ] & {
    stake: BigNumber;
    unstakeDelay: BigNumber;
    withdrawTime: BigNumber;
    abandonedTime: BigNumber;
    keepaliveTime: BigNumber;
    token: string;
    owner: string;
  };

  export type AbandonedRelayServerConfigStruct = {
    devAddress: string;
    abandonmentDelay: BigNumberish;
    escheatmentDelay: BigNumberish;
  };

  export type AbandonedRelayServerConfigStructOutput = [
    string,
    BigNumber,
    BigNumber
  ] & {
    devAddress: string;
    abandonmentDelay: BigNumber;
    escheatmentDelay: BigNumber;
  };
}

export interface IStakeManagerInterface extends utils.Interface {
  functions: {
    "supportsInterface(bytes4)": FunctionFragment;
    "setRelayManagerOwner(address)": FunctionFragment;
    "stakeForRelayManager(address,address,uint256,uint256)": FunctionFragment;
    "unlockStake(address)": FunctionFragment;
    "withdrawStake(address)": FunctionFragment;
    "authorizeHubByOwner(address,address)": FunctionFragment;
    "authorizeHubByManager(address)": FunctionFragment;
    "unauthorizeHubByOwner(address,address)": FunctionFragment;
    "unauthorizeHubByManager(address)": FunctionFragment;
    "penalizeRelayManager(address,address,uint256)": FunctionFragment;
    "markRelayAbandoned(address)": FunctionFragment;
    "escheatAbandonedRelayStake(address)": FunctionFragment;
    "updateRelayKeepaliveTime(address)": FunctionFragment;
    "isRelayEscheatable(address)": FunctionFragment;
    "getStakeInfo(address)": FunctionFragment;
    "getMaxUnstakeDelay()": FunctionFragment;
    "setBurnAddress(address)": FunctionFragment;
    "getBurnAddress()": FunctionFragment;
    "setDevAddress(address)": FunctionFragment;
    "getAbandonedRelayServerConfig()": FunctionFragment;
    "getCreationBlock()": FunctionFragment;
    "versionSM()": FunctionFragment;
  };

  getFunction(
    nameOrSignatureOrTopic:
      | "supportsInterface"
      | "setRelayManagerOwner"
      | "stakeForRelayManager"
      | "unlockStake"
      | "withdrawStake"
      | "authorizeHubByOwner"
      | "authorizeHubByManager"
      | "unauthorizeHubByOwner"
      | "unauthorizeHubByManager"
      | "penalizeRelayManager"
      | "markRelayAbandoned"
      | "escheatAbandonedRelayStake"
      | "updateRelayKeepaliveTime"
      | "isRelayEscheatable"
      | "getStakeInfo"
      | "getMaxUnstakeDelay"
      | "setBurnAddress"
      | "getBurnAddress"
      | "setDevAddress"
      | "getAbandonedRelayServerConfig"
      | "getCreationBlock"
      | "versionSM"
  ): FunctionFragment;

  encodeFunctionData(
    functionFragment: "supportsInterface",
    values: [BytesLike]
  ): string;
  encodeFunctionData(
    functionFragment: "setRelayManagerOwner",
    values: [string]
  ): string;
  encodeFunctionData(
    functionFragment: "stakeForRelayManager",
    values: [string, string, BigNumberish, BigNumberish]
  ): string;
  encodeFunctionData(functionFragment: "unlockStake", values: [string]): string;
  encodeFunctionData(
    functionFragment: "withdrawStake",
    values: [string]
  ): string;
  encodeFunctionData(
    functionFragment: "authorizeHubByOwner",
    values: [string, string]
  ): string;
  encodeFunctionData(
    functionFragment: "authorizeHubByManager",
    values: [string]
  ): string;
  encodeFunctionData(
    functionFragment: "unauthorizeHubByOwner",
    values: [string, string]
  ): string;
  encodeFunctionData(
    functionFragment: "unauthorizeHubByManager",
    values: [string]
  ): string;
  encodeFunctionData(
    functionFragment: "penalizeRelayManager",
    values: [string, string, BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "markRelayAbandoned",
    values: [string]
  ): string;
  encodeFunctionData(
    functionFragment: "escheatAbandonedRelayStake",
    values: [string]
  ): string;
  encodeFunctionData(
    functionFragment: "updateRelayKeepaliveTime",
    values: [string]
  ): string;
  encodeFunctionData(
    functionFragment: "isRelayEscheatable",
    values: [string]
  ): string;
  encodeFunctionData(
    functionFragment: "getStakeInfo",
    values: [string]
  ): string;
  encodeFunctionData(
    functionFragment: "getMaxUnstakeDelay",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "setBurnAddress",
    values: [string]
  ): string;
  encodeFunctionData(
    functionFragment: "getBurnAddress",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "setDevAddress",
    values: [string]
  ): string;
  encodeFunctionData(
    functionFragment: "getAbandonedRelayServerConfig",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "getCreationBlock",
    values?: undefined
  ): string;
  encodeFunctionData(functionFragment: "versionSM", values?: undefined): string;

  decodeFunctionResult(
    functionFragment: "supportsInterface",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "setRelayManagerOwner",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "stakeForRelayManager",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "unlockStake",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "withdrawStake",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "authorizeHubByOwner",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "authorizeHubByManager",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "unauthorizeHubByOwner",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "unauthorizeHubByManager",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "penalizeRelayManager",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "markRelayAbandoned",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "escheatAbandonedRelayStake",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "updateRelayKeepaliveTime",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "isRelayEscheatable",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getStakeInfo",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getMaxUnstakeDelay",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "setBurnAddress",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getBurnAddress",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "setDevAddress",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getAbandonedRelayServerConfig",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getCreationBlock",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "versionSM", data: BytesLike): Result;

  events: {
    "AbandonedRelayManagerStakeEscheated(address,address,address,uint256)": EventFragment;
    "BurnAddressSet(address)": EventFragment;
    "DevAddressSet(address)": EventFragment;
    "HubAuthorized(address,address)": EventFragment;
    "HubUnauthorized(address,address,uint256)": EventFragment;
    "OwnerSet(address,address)": EventFragment;
    "RelayServerAbandoned(address,uint256)": EventFragment;
    "RelayServerKeepalive(address,uint256)": EventFragment;
    "StakeAdded(address,address,address,uint256,uint256)": EventFragment;
    "StakePenalized(address,address,address,uint256)": EventFragment;
    "StakeUnlocked(address,address,uint256)": EventFragment;
    "StakeWithdrawn(address,address,address,uint256)": EventFragment;
  };

  getEvent(
    nameOrSignatureOrTopic: "AbandonedRelayManagerStakeEscheated"
  ): EventFragment;
  getEvent(nameOrSignatureOrTopic: "BurnAddressSet"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "DevAddressSet"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "HubAuthorized"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "HubUnauthorized"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "OwnerSet"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "RelayServerAbandoned"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "RelayServerKeepalive"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "StakeAdded"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "StakePenalized"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "StakeUnlocked"): EventFragment;
  getEvent(nameOrSignatureOrTopic: "StakeWithdrawn"): EventFragment;
}

export interface AbandonedRelayManagerStakeEscheatedEventObject {
  relayManager: string;
  owner: string;
  token: string;
  amount: BigNumber;
}
export type AbandonedRelayManagerStakeEscheatedEvent = TypedEvent<
  [string, string, string, BigNumber],
  AbandonedRelayManagerStakeEscheatedEventObject
>;

export type AbandonedRelayManagerStakeEscheatedEventFilter =
  TypedEventFilter<AbandonedRelayManagerStakeEscheatedEvent>;

export interface BurnAddressSetEventObject {
  burnAddress: string;
}
export type BurnAddressSetEvent = TypedEvent<
  [string],
  BurnAddressSetEventObject
>;

export type BurnAddressSetEventFilter = TypedEventFilter<BurnAddressSetEvent>;

export interface DevAddressSetEventObject {
  devAddress: string;
}
export type DevAddressSetEvent = TypedEvent<[string], DevAddressSetEventObject>;

export type DevAddressSetEventFilter = TypedEventFilter<DevAddressSetEvent>;

export interface HubAuthorizedEventObject {
  relayManager: string;
  relayHub: string;
}
export type HubAuthorizedEvent = TypedEvent<
  [string, string],
  HubAuthorizedEventObject
>;

export type HubAuthorizedEventFilter = TypedEventFilter<HubAuthorizedEvent>;

export interface HubUnauthorizedEventObject {
  relayManager: string;
  relayHub: string;
  removalTime: BigNumber;
}
export type HubUnauthorizedEvent = TypedEvent<
  [string, string, BigNumber],
  HubUnauthorizedEventObject
>;

export type HubUnauthorizedEventFilter = TypedEventFilter<HubUnauthorizedEvent>;

export interface OwnerSetEventObject {
  relayManager: string;
  owner: string;
}
export type OwnerSetEvent = TypedEvent<[string, string], OwnerSetEventObject>;

export type OwnerSetEventFilter = TypedEventFilter<OwnerSetEvent>;

export interface RelayServerAbandonedEventObject {
  relayManager: string;
  abandonedTime: BigNumber;
}
export type RelayServerAbandonedEvent = TypedEvent<
  [string, BigNumber],
  RelayServerAbandonedEventObject
>;

export type RelayServerAbandonedEventFilter =
  TypedEventFilter<RelayServerAbandonedEvent>;

export interface RelayServerKeepaliveEventObject {
  relayManager: string;
  keepaliveTime: BigNumber;
}
export type RelayServerKeepaliveEvent = TypedEvent<
  [string, BigNumber],
  RelayServerKeepaliveEventObject
>;

export type RelayServerKeepaliveEventFilter =
  TypedEventFilter<RelayServerKeepaliveEvent>;

export interface StakeAddedEventObject {
  relayManager: string;
  owner: string;
  token: string;
  stake: BigNumber;
  unstakeDelay: BigNumber;
}
export type StakeAddedEvent = TypedEvent<
  [string, string, string, BigNumber, BigNumber],
  StakeAddedEventObject
>;

export type StakeAddedEventFilter = TypedEventFilter<StakeAddedEvent>;

export interface StakePenalizedEventObject {
  relayManager: string;
  beneficiary: string;
  token: string;
  reward: BigNumber;
}
export type StakePenalizedEvent = TypedEvent<
  [string, string, string, BigNumber],
  StakePenalizedEventObject
>;

export type StakePenalizedEventFilter = TypedEventFilter<StakePenalizedEvent>;

export interface StakeUnlockedEventObject {
  relayManager: string;
  owner: string;
  withdrawTime: BigNumber;
}
export type StakeUnlockedEvent = TypedEvent<
  [string, string, BigNumber],
  StakeUnlockedEventObject
>;

export type StakeUnlockedEventFilter = TypedEventFilter<StakeUnlockedEvent>;

export interface StakeWithdrawnEventObject {
  relayManager: string;
  owner: string;
  token: string;
  amount: BigNumber;
}
export type StakeWithdrawnEvent = TypedEvent<
  [string, string, string, BigNumber],
  StakeWithdrawnEventObject
>;

export type StakeWithdrawnEventFilter = TypedEventFilter<StakeWithdrawnEvent>;

export interface IStakeManager extends BaseContract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: IStakeManagerInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    /**
     * Returns true if this contract implements the interface defined by `interfaceId`. See the corresponding https://eips.ethereum.org/EIPS/eip-165#how-interfaces-are-identified[EIP section] to learn more about how these ids are created. This function call must use less than 30 000 gas.
     */
    supportsInterface(
      interfaceId: BytesLike,
      overrides?: CallOverrides
    ): Promise<[boolean]>;

    /**
     * Set the owner of a Relay Manager. Called only by the RelayManager itself. Note that owners cannot transfer ownership - if the entry already exists, reverts.
     * @param owner - owner of the relay (as configured off-chain)
     */
    setRelayManagerOwner(
      owner: string,
      overrides?: Overrides & { from?: string }
    ): Promise<ContractTransaction>;

    /**
     * Put a stake for a relayManager and set its unstake delay. Only the owner can call this function. If the entry does not exist, reverts. The owner must give allowance of the ERC-20 token to the StakeManager before calling this method. It is the RelayHub who has a configurable list of minimum stakes per token. StakeManager accepts all tokens.
     * @param amount The amount of tokens to be taken from the relayOwner and locked in the StakeManager as a stake
     * @param relayManager The address that represents a stake entry and controls relay registrations on relay hubs
     * @param token The address of an ERC-20 token that is used by the relayManager as a stake
     * @param unstakeDelay The number of seconds to elapse before an owner can retrieve the stake after calling `unlock`
     */
    stakeForRelayManager(
      token: string,
      relayManager: string,
      unstakeDelay: BigNumberish,
      amount: BigNumberish,
      overrides?: Overrides & { from?: string }
    ): Promise<ContractTransaction>;

    /**
     * Schedule the unlocking of the stake. The `unstakeDelay` must pass before owner can call `withdrawStake`.
     * @param relayManager The address of a Relay Manager whose stake is to be unlocked.
     */
    unlockStake(
      relayManager: string,
      overrides?: Overrides & { from?: string }
    ): Promise<ContractTransaction>;

    /**
     * Withdraw the unlocked stake.
     * @param relayManager The address of a Relay Manager whose stake is to be withdrawn.
     */
    withdrawStake(
      relayManager: string,
      overrides?: Overrides & { from?: string }
    ): Promise<ContractTransaction>;

    /**
     * Add the `RelayHub` to a list of authorized by this Relay Manager. This allows the RelayHub to penalize this Relay Manager. The `RelayHub` cannot trust a Relay it cannot penalize.
     * @param relayHub The address of a `RelayHub` to be authorized.
     * @param relayManager The address of a Relay Manager whose stake is to be authorized for the new `RelayHub`.
     */
    authorizeHubByOwner(
      relayManager: string,
      relayHub: string,
      overrides?: Overrides & { from?: string }
    ): Promise<ContractTransaction>;

    /**
     * Same as `authorizeHubByOwner` but can be called by the RelayManager itself.
     */
    authorizeHubByManager(
      relayHub: string,
      overrides?: Overrides & { from?: string }
    ): Promise<ContractTransaction>;

    /**
     * Remove the `RelayHub` from a list of authorized by this Relay Manager.
     * @param relayHub The address of a `RelayHub` to be unauthorized.
     * @param relayManager The address of a Relay Manager.
     */
    unauthorizeHubByOwner(
      relayManager: string,
      relayHub: string,
      overrides?: Overrides & { from?: string }
    ): Promise<ContractTransaction>;

    /**
     * Same as `unauthorizeHubByOwner` but can be called by the RelayManager itself.
     */
    unauthorizeHubByManager(
      relayHub: string,
      overrides?: Overrides & { from?: string }
    ): Promise<ContractTransaction>;

    /**
     * Slash the stake of the relay relayManager. In order to prevent stake kidnapping, burns part of stake on the way.
     * @param amount A total amount of penalty to be withdrawn from stake.
     * @param beneficiary The address that receives part of the penalty amount.
     * @param relayManager The address of a Relay Manager to be penalized.
     */
    penalizeRelayManager(
      relayManager: string,
      beneficiary: string,
      amount: BigNumberish,
      overrides?: Overrides & { from?: string }
    ): Promise<ContractTransaction>;

    /**
     * Allows the contract owner to set the given `relayManager` as abandoned after a configurable delay. Its entire stake and balance will be taken from a relay if it does not respond to being marked as abandoned.
     */
    markRelayAbandoned(
      relayManager: string,
      overrides?: Overrides & { from?: string }
    ): Promise<ContractTransaction>;

    /**
     * If more than `abandonmentDelay` has passed since the last Keepalive transaction, and relay manager has been marked as abandoned, and after that more that `escheatmentDelay` have passed, entire stake and balance will be taken from this relay.
     */
    escheatAbandonedRelayStake(
      relayManager: string,
      overrides?: Overrides & { from?: string }
    ): Promise<ContractTransaction>;

    /**
     * Sets a new `keepaliveTime` for the given `relayManager`, preventing it from being marked as abandoned. Can be called by an authorized `RelayHub` or by the `relayOwner` address.
     */
    updateRelayKeepaliveTime(
      relayManager: string,
      overrides?: Overrides & { from?: string }
    ): Promise<ContractTransaction>;

    /**
     * Check if the Relay Manager can be considered abandoned or not. Returns true if the stake's abandonment time is in the past including the escheatment delay, false otherwise.
     */
    isRelayEscheatable(
      relayManager: string,
      overrides?: CallOverrides
    ): Promise<[boolean]>;

    /**
     * Get the stake details information for the given Relay Manager.
     * @param relayManager The address of a Relay Manager.
     */
    getStakeInfo(
      relayManager: string,
      overrides?: CallOverrides
    ): Promise<
      [IStakeManager.StakeInfoStructOutput, boolean] & {
        stakeInfo: IStakeManager.StakeInfoStructOutput;
        isSenderAuthorizedHub: boolean;
      }
    >;

    /**
     */
    getMaxUnstakeDelay(overrides?: CallOverrides): Promise<[BigNumber]>;

    /**
     * Change the address that will receive the 'burned' part of the penalized stake. This is done to prevent malicious Relay Server from penalizing itself and breaking even.
     */
    setBurnAddress(
      _burnAddress: string,
      overrides?: Overrides & { from?: string }
    ): Promise<ContractTransaction>;

    /**
     */
    getBurnAddress(overrides?: CallOverrides): Promise<[string]>;

    /**
     * Change the address that will receive the 'abandoned' stake. This is done to prevent Relay Servers that lost their keys from losing access to funds.
     */
    setDevAddress(
      _burnAddress: string,
      overrides?: Overrides & { from?: string }
    ): Promise<ContractTransaction>;

    /**
     */
    getAbandonedRelayServerConfig(
      overrides?: CallOverrides
    ): Promise<[IStakeManager.AbandonedRelayServerConfigStructOutput]>;

    /**
     */
    getCreationBlock(overrides?: CallOverrides): Promise<[BigNumber]>;

    /**
     */
    versionSM(overrides?: CallOverrides): Promise<[string]>;
  };

  /**
   * Returns true if this contract implements the interface defined by `interfaceId`. See the corresponding https://eips.ethereum.org/EIPS/eip-165#how-interfaces-are-identified[EIP section] to learn more about how these ids are created. This function call must use less than 30 000 gas.
   */
  supportsInterface(
    interfaceId: BytesLike,
    overrides?: CallOverrides
  ): Promise<boolean>;

  /**
   * Set the owner of a Relay Manager. Called only by the RelayManager itself. Note that owners cannot transfer ownership - if the entry already exists, reverts.
   * @param owner - owner of the relay (as configured off-chain)
   */
  setRelayManagerOwner(
    owner: string,
    overrides?: Overrides & { from?: string }
  ): Promise<ContractTransaction>;

  /**
   * Put a stake for a relayManager and set its unstake delay. Only the owner can call this function. If the entry does not exist, reverts. The owner must give allowance of the ERC-20 token to the StakeManager before calling this method. It is the RelayHub who has a configurable list of minimum stakes per token. StakeManager accepts all tokens.
   * @param amount The amount of tokens to be taken from the relayOwner and locked in the StakeManager as a stake
   * @param relayManager The address that represents a stake entry and controls relay registrations on relay hubs
   * @param token The address of an ERC-20 token that is used by the relayManager as a stake
   * @param unstakeDelay The number of seconds to elapse before an owner can retrieve the stake after calling `unlock`
   */
  stakeForRelayManager(
    token: string,
    relayManager: string,
    unstakeDelay: BigNumberish,
    amount: BigNumberish,
    overrides?: Overrides & { from?: string }
  ): Promise<ContractTransaction>;

  /**
   * Schedule the unlocking of the stake. The `unstakeDelay` must pass before owner can call `withdrawStake`.
   * @param relayManager The address of a Relay Manager whose stake is to be unlocked.
   */
  unlockStake(
    relayManager: string,
    overrides?: Overrides & { from?: string }
  ): Promise<ContractTransaction>;

  /**
   * Withdraw the unlocked stake.
   * @param relayManager The address of a Relay Manager whose stake is to be withdrawn.
   */
  withdrawStake(
    relayManager: string,
    overrides?: Overrides & { from?: string }
  ): Promise<ContractTransaction>;

  /**
   * Add the `RelayHub` to a list of authorized by this Relay Manager. This allows the RelayHub to penalize this Relay Manager. The `RelayHub` cannot trust a Relay it cannot penalize.
   * @param relayHub The address of a `RelayHub` to be authorized.
   * @param relayManager The address of a Relay Manager whose stake is to be authorized for the new `RelayHub`.
   */
  authorizeHubByOwner(
    relayManager: string,
    relayHub: string,
    overrides?: Overrides & { from?: string }
  ): Promise<ContractTransaction>;

  /**
   * Same as `authorizeHubByOwner` but can be called by the RelayManager itself.
   */
  authorizeHubByManager(
    relayHub: string,
    overrides?: Overrides & { from?: string }
  ): Promise<ContractTransaction>;

  /**
   * Remove the `RelayHub` from a list of authorized by this Relay Manager.
   * @param relayHub The address of a `RelayHub` to be unauthorized.
   * @param relayManager The address of a Relay Manager.
   */
  unauthorizeHubByOwner(
    relayManager: string,
    relayHub: string,
    overrides?: Overrides & { from?: string }
  ): Promise<ContractTransaction>;

  /**
   * Same as `unauthorizeHubByOwner` but can be called by the RelayManager itself.
   */
  unauthorizeHubByManager(
    relayHub: string,
    overrides?: Overrides & { from?: string }
  ): Promise<ContractTransaction>;

  /**
   * Slash the stake of the relay relayManager. In order to prevent stake kidnapping, burns part of stake on the way.
   * @param amount A total amount of penalty to be withdrawn from stake.
   * @param beneficiary The address that receives part of the penalty amount.
   * @param relayManager The address of a Relay Manager to be penalized.
   */
  penalizeRelayManager(
    relayManager: string,
    beneficiary: string,
    amount: BigNumberish,
    overrides?: Overrides & { from?: string }
  ): Promise<ContractTransaction>;

  /**
   * Allows the contract owner to set the given `relayManager` as abandoned after a configurable delay. Its entire stake and balance will be taken from a relay if it does not respond to being marked as abandoned.
   */
  markRelayAbandoned(
    relayManager: string,
    overrides?: Overrides & { from?: string }
  ): Promise<ContractTransaction>;

  /**
   * If more than `abandonmentDelay` has passed since the last Keepalive transaction, and relay manager has been marked as abandoned, and after that more that `escheatmentDelay` have passed, entire stake and balance will be taken from this relay.
   */
  escheatAbandonedRelayStake(
    relayManager: string,
    overrides?: Overrides & { from?: string }
  ): Promise<ContractTransaction>;

  /**
   * Sets a new `keepaliveTime` for the given `relayManager`, preventing it from being marked as abandoned. Can be called by an authorized `RelayHub` or by the `relayOwner` address.
   */
  updateRelayKeepaliveTime(
    relayManager: string,
    overrides?: Overrides & { from?: string }
  ): Promise<ContractTransaction>;

  /**
   * Check if the Relay Manager can be considered abandoned or not. Returns true if the stake's abandonment time is in the past including the escheatment delay, false otherwise.
   */
  isRelayEscheatable(
    relayManager: string,
    overrides?: CallOverrides
  ): Promise<boolean>;

  /**
   * Get the stake details information for the given Relay Manager.
   * @param relayManager The address of a Relay Manager.
   */
  getStakeInfo(
    relayManager: string,
    overrides?: CallOverrides
  ): Promise<
    [IStakeManager.StakeInfoStructOutput, boolean] & {
      stakeInfo: IStakeManager.StakeInfoStructOutput;
      isSenderAuthorizedHub: boolean;
    }
  >;

  /**
   */
  getMaxUnstakeDelay(overrides?: CallOverrides): Promise<BigNumber>;

  /**
   * Change the address that will receive the 'burned' part of the penalized stake. This is done to prevent malicious Relay Server from penalizing itself and breaking even.
   */
  setBurnAddress(
    _burnAddress: string,
    overrides?: Overrides & { from?: string }
  ): Promise<ContractTransaction>;

  /**
   */
  getBurnAddress(overrides?: CallOverrides): Promise<string>;

  /**
   * Change the address that will receive the 'abandoned' stake. This is done to prevent Relay Servers that lost their keys from losing access to funds.
   */
  setDevAddress(
    _burnAddress: string,
    overrides?: Overrides & { from?: string }
  ): Promise<ContractTransaction>;

  /**
   */
  getAbandonedRelayServerConfig(
    overrides?: CallOverrides
  ): Promise<IStakeManager.AbandonedRelayServerConfigStructOutput>;

  /**
   */
  getCreationBlock(overrides?: CallOverrides): Promise<BigNumber>;

  /**
   */
  versionSM(overrides?: CallOverrides): Promise<string>;

  callStatic: {
    /**
     * Returns true if this contract implements the interface defined by `interfaceId`. See the corresponding https://eips.ethereum.org/EIPS/eip-165#how-interfaces-are-identified[EIP section] to learn more about how these ids are created. This function call must use less than 30 000 gas.
     */
    supportsInterface(
      interfaceId: BytesLike,
      overrides?: CallOverrides
    ): Promise<boolean>;

    /**
     * Set the owner of a Relay Manager. Called only by the RelayManager itself. Note that owners cannot transfer ownership - if the entry already exists, reverts.
     * @param owner - owner of the relay (as configured off-chain)
     */
    setRelayManagerOwner(
      owner: string,
      overrides?: CallOverrides
    ): Promise<void>;

    /**
     * Put a stake for a relayManager and set its unstake delay. Only the owner can call this function. If the entry does not exist, reverts. The owner must give allowance of the ERC-20 token to the StakeManager before calling this method. It is the RelayHub who has a configurable list of minimum stakes per token. StakeManager accepts all tokens.
     * @param amount The amount of tokens to be taken from the relayOwner and locked in the StakeManager as a stake
     * @param relayManager The address that represents a stake entry and controls relay registrations on relay hubs
     * @param token The address of an ERC-20 token that is used by the relayManager as a stake
     * @param unstakeDelay The number of seconds to elapse before an owner can retrieve the stake after calling `unlock`
     */
    stakeForRelayManager(
      token: string,
      relayManager: string,
      unstakeDelay: BigNumberish,
      amount: BigNumberish,
      overrides?: CallOverrides
    ): Promise<void>;

    /**
     * Schedule the unlocking of the stake. The `unstakeDelay` must pass before owner can call `withdrawStake`.
     * @param relayManager The address of a Relay Manager whose stake is to be unlocked.
     */
    unlockStake(relayManager: string, overrides?: CallOverrides): Promise<void>;

    /**
     * Withdraw the unlocked stake.
     * @param relayManager The address of a Relay Manager whose stake is to be withdrawn.
     */
    withdrawStake(
      relayManager: string,
      overrides?: CallOverrides
    ): Promise<void>;

    /**
     * Add the `RelayHub` to a list of authorized by this Relay Manager. This allows the RelayHub to penalize this Relay Manager. The `RelayHub` cannot trust a Relay it cannot penalize.
     * @param relayHub The address of a `RelayHub` to be authorized.
     * @param relayManager The address of a Relay Manager whose stake is to be authorized for the new `RelayHub`.
     */
    authorizeHubByOwner(
      relayManager: string,
      relayHub: string,
      overrides?: CallOverrides
    ): Promise<void>;

    /**
     * Same as `authorizeHubByOwner` but can be called by the RelayManager itself.
     */
    authorizeHubByManager(
      relayHub: string,
      overrides?: CallOverrides
    ): Promise<void>;

    /**
     * Remove the `RelayHub` from a list of authorized by this Relay Manager.
     * @param relayHub The address of a `RelayHub` to be unauthorized.
     * @param relayManager The address of a Relay Manager.
     */
    unauthorizeHubByOwner(
      relayManager: string,
      relayHub: string,
      overrides?: CallOverrides
    ): Promise<void>;

    /**
     * Same as `unauthorizeHubByOwner` but can be called by the RelayManager itself.
     */
    unauthorizeHubByManager(
      relayHub: string,
      overrides?: CallOverrides
    ): Promise<void>;

    /**
     * Slash the stake of the relay relayManager. In order to prevent stake kidnapping, burns part of stake on the way.
     * @param amount A total amount of penalty to be withdrawn from stake.
     * @param beneficiary The address that receives part of the penalty amount.
     * @param relayManager The address of a Relay Manager to be penalized.
     */
    penalizeRelayManager(
      relayManager: string,
      beneficiary: string,
      amount: BigNumberish,
      overrides?: CallOverrides
    ): Promise<void>;

    /**
     * Allows the contract owner to set the given `relayManager` as abandoned after a configurable delay. Its entire stake and balance will be taken from a relay if it does not respond to being marked as abandoned.
     */
    markRelayAbandoned(
      relayManager: string,
      overrides?: CallOverrides
    ): Promise<void>;

    /**
     * If more than `abandonmentDelay` has passed since the last Keepalive transaction, and relay manager has been marked as abandoned, and after that more that `escheatmentDelay` have passed, entire stake and balance will be taken from this relay.
     */
    escheatAbandonedRelayStake(
      relayManager: string,
      overrides?: CallOverrides
    ): Promise<void>;

    /**
     * Sets a new `keepaliveTime` for the given `relayManager`, preventing it from being marked as abandoned. Can be called by an authorized `RelayHub` or by the `relayOwner` address.
     */
    updateRelayKeepaliveTime(
      relayManager: string,
      overrides?: CallOverrides
    ): Promise<void>;

    /**
     * Check if the Relay Manager can be considered abandoned or not. Returns true if the stake's abandonment time is in the past including the escheatment delay, false otherwise.
     */
    isRelayEscheatable(
      relayManager: string,
      overrides?: CallOverrides
    ): Promise<boolean>;

    /**
     * Get the stake details information for the given Relay Manager.
     * @param relayManager The address of a Relay Manager.
     */
    getStakeInfo(
      relayManager: string,
      overrides?: CallOverrides
    ): Promise<
      [IStakeManager.StakeInfoStructOutput, boolean] & {
        stakeInfo: IStakeManager.StakeInfoStructOutput;
        isSenderAuthorizedHub: boolean;
      }
    >;

    /**
     */
    getMaxUnstakeDelay(overrides?: CallOverrides): Promise<BigNumber>;

    /**
     * Change the address that will receive the 'burned' part of the penalized stake. This is done to prevent malicious Relay Server from penalizing itself and breaking even.
     */
    setBurnAddress(
      _burnAddress: string,
      overrides?: CallOverrides
    ): Promise<void>;

    /**
     */
    getBurnAddress(overrides?: CallOverrides): Promise<string>;

    /**
     * Change the address that will receive the 'abandoned' stake. This is done to prevent Relay Servers that lost their keys from losing access to funds.
     */
    setDevAddress(
      _burnAddress: string,
      overrides?: CallOverrides
    ): Promise<void>;

    /**
     */
    getAbandonedRelayServerConfig(
      overrides?: CallOverrides
    ): Promise<IStakeManager.AbandonedRelayServerConfigStructOutput>;

    /**
     */
    getCreationBlock(overrides?: CallOverrides): Promise<BigNumber>;

    /**
     */
    versionSM(overrides?: CallOverrides): Promise<string>;
  };

  filters: {
    "AbandonedRelayManagerStakeEscheated(address,address,address,uint256)"(
      relayManager?: string | null,
      owner?: string | null,
      token?: null,
      amount?: null
    ): AbandonedRelayManagerStakeEscheatedEventFilter;
    AbandonedRelayManagerStakeEscheated(
      relayManager?: string | null,
      owner?: string | null,
      token?: null,
      amount?: null
    ): AbandonedRelayManagerStakeEscheatedEventFilter;

    "BurnAddressSet(address)"(
      burnAddress?: string | null
    ): BurnAddressSetEventFilter;
    BurnAddressSet(burnAddress?: string | null): BurnAddressSetEventFilter;

    "DevAddressSet(address)"(
      devAddress?: string | null
    ): DevAddressSetEventFilter;
    DevAddressSet(devAddress?: string | null): DevAddressSetEventFilter;

    "HubAuthorized(address,address)"(
      relayManager?: string | null,
      relayHub?: string | null
    ): HubAuthorizedEventFilter;
    HubAuthorized(
      relayManager?: string | null,
      relayHub?: string | null
    ): HubAuthorizedEventFilter;

    "HubUnauthorized(address,address,uint256)"(
      relayManager?: string | null,
      relayHub?: string | null,
      removalTime?: null
    ): HubUnauthorizedEventFilter;
    HubUnauthorized(
      relayManager?: string | null,
      relayHub?: string | null,
      removalTime?: null
    ): HubUnauthorizedEventFilter;

    "OwnerSet(address,address)"(
      relayManager?: string | null,
      owner?: string | null
    ): OwnerSetEventFilter;
    OwnerSet(
      relayManager?: string | null,
      owner?: string | null
    ): OwnerSetEventFilter;

    "RelayServerAbandoned(address,uint256)"(
      relayManager?: string | null,
      abandonedTime?: null
    ): RelayServerAbandonedEventFilter;
    RelayServerAbandoned(
      relayManager?: string | null,
      abandonedTime?: null
    ): RelayServerAbandonedEventFilter;

    "RelayServerKeepalive(address,uint256)"(
      relayManager?: string | null,
      keepaliveTime?: null
    ): RelayServerKeepaliveEventFilter;
    RelayServerKeepalive(
      relayManager?: string | null,
      keepaliveTime?: null
    ): RelayServerKeepaliveEventFilter;

    "StakeAdded(address,address,address,uint256,uint256)"(
      relayManager?: string | null,
      owner?: string | null,
      token?: null,
      stake?: null,
      unstakeDelay?: null
    ): StakeAddedEventFilter;
    StakeAdded(
      relayManager?: string | null,
      owner?: string | null,
      token?: null,
      stake?: null,
      unstakeDelay?: null
    ): StakeAddedEventFilter;

    "StakePenalized(address,address,address,uint256)"(
      relayManager?: string | null,
      beneficiary?: string | null,
      token?: null,
      reward?: null
    ): StakePenalizedEventFilter;
    StakePenalized(
      relayManager?: string | null,
      beneficiary?: string | null,
      token?: null,
      reward?: null
    ): StakePenalizedEventFilter;

    "StakeUnlocked(address,address,uint256)"(
      relayManager?: string | null,
      owner?: string | null,
      withdrawTime?: null
    ): StakeUnlockedEventFilter;
    StakeUnlocked(
      relayManager?: string | null,
      owner?: string | null,
      withdrawTime?: null
    ): StakeUnlockedEventFilter;

    "StakeWithdrawn(address,address,address,uint256)"(
      relayManager?: string | null,
      owner?: string | null,
      token?: null,
      amount?: null
    ): StakeWithdrawnEventFilter;
    StakeWithdrawn(
      relayManager?: string | null,
      owner?: string | null,
      token?: null,
      amount?: null
    ): StakeWithdrawnEventFilter;
  };

  estimateGas: {
    /**
     * Returns true if this contract implements the interface defined by `interfaceId`. See the corresponding https://eips.ethereum.org/EIPS/eip-165#how-interfaces-are-identified[EIP section] to learn more about how these ids are created. This function call must use less than 30 000 gas.
     */
    supportsInterface(
      interfaceId: BytesLike,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    /**
     * Set the owner of a Relay Manager. Called only by the RelayManager itself. Note that owners cannot transfer ownership - if the entry already exists, reverts.
     * @param owner - owner of the relay (as configured off-chain)
     */
    setRelayManagerOwner(
      owner: string,
      overrides?: Overrides & { from?: string }
    ): Promise<BigNumber>;

    /**
     * Put a stake for a relayManager and set its unstake delay. Only the owner can call this function. If the entry does not exist, reverts. The owner must give allowance of the ERC-20 token to the StakeManager before calling this method. It is the RelayHub who has a configurable list of minimum stakes per token. StakeManager accepts all tokens.
     * @param amount The amount of tokens to be taken from the relayOwner and locked in the StakeManager as a stake
     * @param relayManager The address that represents a stake entry and controls relay registrations on relay hubs
     * @param token The address of an ERC-20 token that is used by the relayManager as a stake
     * @param unstakeDelay The number of seconds to elapse before an owner can retrieve the stake after calling `unlock`
     */
    stakeForRelayManager(
      token: string,
      relayManager: string,
      unstakeDelay: BigNumberish,
      amount: BigNumberish,
      overrides?: Overrides & { from?: string }
    ): Promise<BigNumber>;

    /**
     * Schedule the unlocking of the stake. The `unstakeDelay` must pass before owner can call `withdrawStake`.
     * @param relayManager The address of a Relay Manager whose stake is to be unlocked.
     */
    unlockStake(
      relayManager: string,
      overrides?: Overrides & { from?: string }
    ): Promise<BigNumber>;

    /**
     * Withdraw the unlocked stake.
     * @param relayManager The address of a Relay Manager whose stake is to be withdrawn.
     */
    withdrawStake(
      relayManager: string,
      overrides?: Overrides & { from?: string }
    ): Promise<BigNumber>;

    /**
     * Add the `RelayHub` to a list of authorized by this Relay Manager. This allows the RelayHub to penalize this Relay Manager. The `RelayHub` cannot trust a Relay it cannot penalize.
     * @param relayHub The address of a `RelayHub` to be authorized.
     * @param relayManager The address of a Relay Manager whose stake is to be authorized for the new `RelayHub`.
     */
    authorizeHubByOwner(
      relayManager: string,
      relayHub: string,
      overrides?: Overrides & { from?: string }
    ): Promise<BigNumber>;

    /**
     * Same as `authorizeHubByOwner` but can be called by the RelayManager itself.
     */
    authorizeHubByManager(
      relayHub: string,
      overrides?: Overrides & { from?: string }
    ): Promise<BigNumber>;

    /**
     * Remove the `RelayHub` from a list of authorized by this Relay Manager.
     * @param relayHub The address of a `RelayHub` to be unauthorized.
     * @param relayManager The address of a Relay Manager.
     */
    unauthorizeHubByOwner(
      relayManager: string,
      relayHub: string,
      overrides?: Overrides & { from?: string }
    ): Promise<BigNumber>;

    /**
     * Same as `unauthorizeHubByOwner` but can be called by the RelayManager itself.
     */
    unauthorizeHubByManager(
      relayHub: string,
      overrides?: Overrides & { from?: string }
    ): Promise<BigNumber>;

    /**
     * Slash the stake of the relay relayManager. In order to prevent stake kidnapping, burns part of stake on the way.
     * @param amount A total amount of penalty to be withdrawn from stake.
     * @param beneficiary The address that receives part of the penalty amount.
     * @param relayManager The address of a Relay Manager to be penalized.
     */
    penalizeRelayManager(
      relayManager: string,
      beneficiary: string,
      amount: BigNumberish,
      overrides?: Overrides & { from?: string }
    ): Promise<BigNumber>;

    /**
     * Allows the contract owner to set the given `relayManager` as abandoned after a configurable delay. Its entire stake and balance will be taken from a relay if it does not respond to being marked as abandoned.
     */
    markRelayAbandoned(
      relayManager: string,
      overrides?: Overrides & { from?: string }
    ): Promise<BigNumber>;

    /**
     * If more than `abandonmentDelay` has passed since the last Keepalive transaction, and relay manager has been marked as abandoned, and after that more that `escheatmentDelay` have passed, entire stake and balance will be taken from this relay.
     */
    escheatAbandonedRelayStake(
      relayManager: string,
      overrides?: Overrides & { from?: string }
    ): Promise<BigNumber>;

    /**
     * Sets a new `keepaliveTime` for the given `relayManager`, preventing it from being marked as abandoned. Can be called by an authorized `RelayHub` or by the `relayOwner` address.
     */
    updateRelayKeepaliveTime(
      relayManager: string,
      overrides?: Overrides & { from?: string }
    ): Promise<BigNumber>;

    /**
     * Check if the Relay Manager can be considered abandoned or not. Returns true if the stake's abandonment time is in the past including the escheatment delay, false otherwise.
     */
    isRelayEscheatable(
      relayManager: string,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    /**
     * Get the stake details information for the given Relay Manager.
     * @param relayManager The address of a Relay Manager.
     */
    getStakeInfo(
      relayManager: string,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    /**
     */
    getMaxUnstakeDelay(overrides?: CallOverrides): Promise<BigNumber>;

    /**
     * Change the address that will receive the 'burned' part of the penalized stake. This is done to prevent malicious Relay Server from penalizing itself and breaking even.
     */
    setBurnAddress(
      _burnAddress: string,
      overrides?: Overrides & { from?: string }
    ): Promise<BigNumber>;

    /**
     */
    getBurnAddress(overrides?: CallOverrides): Promise<BigNumber>;

    /**
     * Change the address that will receive the 'abandoned' stake. This is done to prevent Relay Servers that lost their keys from losing access to funds.
     */
    setDevAddress(
      _burnAddress: string,
      overrides?: Overrides & { from?: string }
    ): Promise<BigNumber>;

    /**
     */
    getAbandonedRelayServerConfig(
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    /**
     */
    getCreationBlock(overrides?: CallOverrides): Promise<BigNumber>;

    /**
     */
    versionSM(overrides?: CallOverrides): Promise<BigNumber>;
  };

  populateTransaction: {
    /**
     * Returns true if this contract implements the interface defined by `interfaceId`. See the corresponding https://eips.ethereum.org/EIPS/eip-165#how-interfaces-are-identified[EIP section] to learn more about how these ids are created. This function call must use less than 30 000 gas.
     */
    supportsInterface(
      interfaceId: BytesLike,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    /**
     * Set the owner of a Relay Manager. Called only by the RelayManager itself. Note that owners cannot transfer ownership - if the entry already exists, reverts.
     * @param owner - owner of the relay (as configured off-chain)
     */
    setRelayManagerOwner(
      owner: string,
      overrides?: Overrides & { from?: string }
    ): Promise<PopulatedTransaction>;

    /**
     * Put a stake for a relayManager and set its unstake delay. Only the owner can call this function. If the entry does not exist, reverts. The owner must give allowance of the ERC-20 token to the StakeManager before calling this method. It is the RelayHub who has a configurable list of minimum stakes per token. StakeManager accepts all tokens.
     * @param amount The amount of tokens to be taken from the relayOwner and locked in the StakeManager as a stake
     * @param relayManager The address that represents a stake entry and controls relay registrations on relay hubs
     * @param token The address of an ERC-20 token that is used by the relayManager as a stake
     * @param unstakeDelay The number of seconds to elapse before an owner can retrieve the stake after calling `unlock`
     */
    stakeForRelayManager(
      token: string,
      relayManager: string,
      unstakeDelay: BigNumberish,
      amount: BigNumberish,
      overrides?: Overrides & { from?: string }
    ): Promise<PopulatedTransaction>;

    /**
     * Schedule the unlocking of the stake. The `unstakeDelay` must pass before owner can call `withdrawStake`.
     * @param relayManager The address of a Relay Manager whose stake is to be unlocked.
     */
    unlockStake(
      relayManager: string,
      overrides?: Overrides & { from?: string }
    ): Promise<PopulatedTransaction>;

    /**
     * Withdraw the unlocked stake.
     * @param relayManager The address of a Relay Manager whose stake is to be withdrawn.
     */
    withdrawStake(
      relayManager: string,
      overrides?: Overrides & { from?: string }
    ): Promise<PopulatedTransaction>;

    /**
     * Add the `RelayHub` to a list of authorized by this Relay Manager. This allows the RelayHub to penalize this Relay Manager. The `RelayHub` cannot trust a Relay it cannot penalize.
     * @param relayHub The address of a `RelayHub` to be authorized.
     * @param relayManager The address of a Relay Manager whose stake is to be authorized for the new `RelayHub`.
     */
    authorizeHubByOwner(
      relayManager: string,
      relayHub: string,
      overrides?: Overrides & { from?: string }
    ): Promise<PopulatedTransaction>;

    /**
     * Same as `authorizeHubByOwner` but can be called by the RelayManager itself.
     */
    authorizeHubByManager(
      relayHub: string,
      overrides?: Overrides & { from?: string }
    ): Promise<PopulatedTransaction>;

    /**
     * Remove the `RelayHub` from a list of authorized by this Relay Manager.
     * @param relayHub The address of a `RelayHub` to be unauthorized.
     * @param relayManager The address of a Relay Manager.
     */
    unauthorizeHubByOwner(
      relayManager: string,
      relayHub: string,
      overrides?: Overrides & { from?: string }
    ): Promise<PopulatedTransaction>;

    /**
     * Same as `unauthorizeHubByOwner` but can be called by the RelayManager itself.
     */
    unauthorizeHubByManager(
      relayHub: string,
      overrides?: Overrides & { from?: string }
    ): Promise<PopulatedTransaction>;

    /**
     * Slash the stake of the relay relayManager. In order to prevent stake kidnapping, burns part of stake on the way.
     * @param amount A total amount of penalty to be withdrawn from stake.
     * @param beneficiary The address that receives part of the penalty amount.
     * @param relayManager The address of a Relay Manager to be penalized.
     */
    penalizeRelayManager(
      relayManager: string,
      beneficiary: string,
      amount: BigNumberish,
      overrides?: Overrides & { from?: string }
    ): Promise<PopulatedTransaction>;

    /**
     * Allows the contract owner to set the given `relayManager` as abandoned after a configurable delay. Its entire stake and balance will be taken from a relay if it does not respond to being marked as abandoned.
     */
    markRelayAbandoned(
      relayManager: string,
      overrides?: Overrides & { from?: string }
    ): Promise<PopulatedTransaction>;

    /**
     * If more than `abandonmentDelay` has passed since the last Keepalive transaction, and relay manager has been marked as abandoned, and after that more that `escheatmentDelay` have passed, entire stake and balance will be taken from this relay.
     */
    escheatAbandonedRelayStake(
      relayManager: string,
      overrides?: Overrides & { from?: string }
    ): Promise<PopulatedTransaction>;

    /**
     * Sets a new `keepaliveTime` for the given `relayManager`, preventing it from being marked as abandoned. Can be called by an authorized `RelayHub` or by the `relayOwner` address.
     */
    updateRelayKeepaliveTime(
      relayManager: string,
      overrides?: Overrides & { from?: string }
    ): Promise<PopulatedTransaction>;

    /**
     * Check if the Relay Manager can be considered abandoned or not. Returns true if the stake's abandonment time is in the past including the escheatment delay, false otherwise.
     */
    isRelayEscheatable(
      relayManager: string,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    /**
     * Get the stake details information for the given Relay Manager.
     * @param relayManager The address of a Relay Manager.
     */
    getStakeInfo(
      relayManager: string,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    /**
     */
    getMaxUnstakeDelay(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    /**
     * Change the address that will receive the 'burned' part of the penalized stake. This is done to prevent malicious Relay Server from penalizing itself and breaking even.
     */
    setBurnAddress(
      _burnAddress: string,
      overrides?: Overrides & { from?: string }
    ): Promise<PopulatedTransaction>;

    /**
     */
    getBurnAddress(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    /**
     * Change the address that will receive the 'abandoned' stake. This is done to prevent Relay Servers that lost their keys from losing access to funds.
     */
    setDevAddress(
      _burnAddress: string,
      overrides?: Overrides & { from?: string }
    ): Promise<PopulatedTransaction>;

    /**
     */
    getAbandonedRelayServerConfig(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    /**
     */
    getCreationBlock(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    /**
     */
    versionSM(overrides?: CallOverrides): Promise<PopulatedTransaction>;
  };
}
