// @opengsn/contracts/dist/types/ethers-contracts/factories/IPenalizer__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IPenalizer__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "address",
                name: "sender",
                type: "address",
            },
            {
                indexed: true,
                internalType: "bytes32",
                name: "commitHash",
                type: "bytes32",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "readyBlockNumber",
                type: "uint256",
            },
        ],
        name: "CommitAdded",
        type: "event",
    },
    {
        inputs: [
            {
                internalType: "bytes4",
                name: "interfaceId",
                type: "bytes4",
            },
        ],
        name: "supportsInterface",
        outputs: [
            {
                internalType: "bool",
                name: "",
                type: "bool",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "bytes32",
                name: "commitHash",
                type: "bytes32",
            },
        ],
        name: "commit",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "bytes",
                name: "unsignedTx1",
                type: "bytes",
            },
            {
                internalType: "bytes",
                name: "signature1",
                type: "bytes",
            },
            {
                internalType: "bytes",
                name: "unsignedTx2",
                type: "bytes",
            },
            {
                internalType: "bytes",
                name: "signature2",
                type: "bytes",
            },
            {
                internalType: "contract IRelayHub",
                name: "hub",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "randomValue",
                type: "uint256",
            },
        ],
        name: "penalizeRepeatedNonce",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "bytes",
                name: "unsignedTx",
                type: "bytes",
            },
            {
                internalType: "bytes",
                name: "signature",
                type: "bytes",
            },
            {
                internalType: "contract IRelayHub",
                name: "hub",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "randomValue",
                type: "uint256",
            },
        ],
        name: "penalizeIllegalTransaction",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [],
        name: "versionPenalizer",
        outputs: [
            {
                internalType: "string",
                name: "",
                type: "string",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "getPenalizeBlockDelay",
        outputs: [
            {
                internalType: "uint256",
                name: "",
                type: "uint256",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "getPenalizeBlockExpiration",
        outputs: [
            {
                internalType: "uint256",
                name: "",
                type: "uint256",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
];
class IPenalizer__factory {
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.IPenalizer__factory = IPenalizer__factory;
IPenalizer__factory.abi = _abi;
